### ISSUE TITLE
Huge memory usage for GPU when using textures

### ISSUE THREAD

**Issue by terhechte** [google/forma#23]
Hey, so I had some fun porting text rendering to Forma using [parley](https://github.com/dfrg/parley):

<img width="624" alt="Screenshot 2022-12-19 at 22 18 50" src="https://user-images.githubusercontent.com/132234/208526000-d5ec0594-0958-4e9a-96d8-5ffe21f4d524.png">

However, when using the GPU, the memory usage is quite high (e.g. for this demo, in release mode, ~400MB on my Mac). When I then start resizing the window, it quickly grows to ~600MB and beyond.

This doesn't happen when using a CPU runner. If I render text without emoji, then I don't see this behaviour. The way emoji are rendered is by rendering the glyph to a small image (16x16 right now) and then rendering this image with a 16x16 Path. Do I need to perform some sort of cleanup when rendering images using Forma with the GPU renderer?

**Comment by terhechte** [comment]
I've uploaded my code [here](https://github.com/terhechte/tted). When enabling the GPU renderer `examples/demo/main.rs:114` the memory usage goes through the roof.

Also, I'm getting a lot of artefacts when enabling clipping (even on the CPU): The first render is ok, but once I compose again (due to changes in the scene) everything but the first layer seems to disappear. I'm probably doing something wrong here `examples/demo/draw.rs:72` enables clipping.

**Comment by dragostis** [comment]
Thanks for reporting this. I'll take a look at the high memory consumption and report back.

**Comment by dragostis** [comment]
About the cropping, I don't think I understand the issue. If you're trying to crop the geometry that lies outside of the screen, this should be something we optimize anyway. See #25.

**Comment by terhechte** [comment]
Thanks for the reply!

> About the cropping, I don't think I understand the issue. If you're trying to crop the geometry that lies outside of the screen, this should be something we optimize anyway. See https://github.com/google/forma/issues/25.

Yeah, that's what I was trying to do by having one "base" layer (the size of the window) and then clipping all additional layers after that. But what's happening is that this renders fine, but when the scene changes (e.g. I zoom or translate the transform) then most layers seem to disappear (at least that's how I interpret it):

https://user-images.githubusercontent.com/132234/208673646-92b652ee-12a8-4b2c-8e5e-a155a37b1a2e.mp4

**Comment by dragostis** [comment]
I merged #29 which should improve performance when zooming in. We can slowly improve on the other issues as well.

**Comment by terhechte** [comment]
Thanks! I'll try it out tomorrow

**Comment by FoundationKen** [comment]
@terhechte How is the parley integration coming?  Did the performance improve with the PR from @dragostis?

**Comment by terhechte** [comment]
@FoundationKen Yeah, performance and memory usage did improve. Then I got distracted by another side project 🤷
